src/char/int_storage.c

```diff
@@ -74,36 +77,41 @@ int inter_storage_fromsql(int account_id, struct storage_data* p)
-		SQL->GetData(inter->sql_handle, 2, &data, NULL); item->amount = atoi(data);
-		SQL->GetData(inter->sql_handle, 3, &data, NULL); item->equip = atoi(data);
-		SQL->GetData(inter->sql_handle, 4, &data, NULL); item->identify = atoi(data);
-		SQL->GetData(inter->sql_handle, 5, &data, NULL); item->refine = atoi(data);
-		SQL->GetData(inter->sql_handle, 6, &data, NULL); item->attribute = atoi(data);
-		SQL->GetData(inter->sql_handle, 7, &data, NULL); item->expire_time = (unsigned int)atoi(data);
-		SQL->GetData(inter->sql_handle, 8, &data, NULL); item->bound = atoi(data);
-		SQL->GetData(inter->sql_handle, 9, &data, NULL); item->unique_id = strtoull(data, NULL, 10);
-		/* Card Slots */
-		for (j = 0; j < MAX_SLOTS; ++j) {
-			SQL->GetData(inter->sql_handle, 10 + j, &data, NULL);
-			item->card[j] = atoi(data);
-		}
-		/* Item Options */
-		for (j = 0; j < MAX_ITEM_OPTIONS; ++j) {
-			SQL->GetData(inter->sql_handle, 10 + MAX_SLOTS + j * 2, &data, NULL);
-			item->option[j].index = atoi(data);
-			SQL->GetData(inter->sql_handle, 11 + MAX_SLOTS + j * 2, &data, NULL);
-			item->option[j].value = atoi(data);
+	if ((num_rows = (int)SQL->NumRows(inter->sql_handle)) != 0) {
+		VECTOR_ENSURE(*p, num_rows, 1);
+
+		for (i = 0; i < MAX_STORAGE && SQL_SUCCESS == SQL->NextRow(inter->sql_handle); ++i) {
+			struct item item = { 0 };
+			SQL->GetData(inter->sql_handle, 0, &data, NULL); item.id = atoi(data);
+			SQL->GetData(inter->sql_handle, 1, &data, NULL); item.nameid = atoi(data);
+			SQL->GetData(inter->sql_handle, 2, &data, NULL); item.amount = atoi(data);
+			SQL->GetData(inter->sql_handle, 3, &data, NULL); item.equip = atoi(data);
+			SQL->GetData(inter->sql_handle, 4, &data, NULL); item.identify = atoi(data);
+			SQL->GetData(inter->sql_handle, 5, &data, NULL); item.refine = atoi(data);
+			SQL->GetData(inter->sql_handle, 6, &data, NULL); item.attribute = atoi(data);
+			SQL->GetData(inter->sql_handle, 7, &data, NULL); item.expire_time = (unsigned int)atoi(data);
+			SQL->GetData(inter->sql_handle, 8, &data, NULL); item.bound = atoi(data);
+			SQL->GetData(inter->sql_handle, 9, &data, NULL); item.unique_id = strtoull(data, NULL, 10);
+			/* Card Slots */
+			for (j = 0; j < MAX_SLOTS; ++j) {
+				SQL->GetData(inter->sql_handle, 10 + j, &data, NULL);
+				item.card[j] = atoi(data);
+			}
+			/* Item Options */
+			for (j = 0; j < MAX_ITEM_OPTIONS; ++j) {
+				SQL->GetData(inter->sql_handle, 10 + MAX_SLOTS + j * 2, &data, NULL);
+				item.option[j].index = atoi(data);
+				SQL->GetData(inter->sql_handle, 11 + MAX_SLOTS + j * 2, &data, NULL);
+				item.option[j].value = atoi(data);
+			}
+			VECTOR_PUSH(*p, item);
 		}
 	}
-	p->storage_amount = i;
+
 	SQL->FreeResult(inter->sql_handle);
 
-	ShowInfo("storage load complete from DB - id: %d (total: %d)\n", account_id, p->storage_amount);
-	return 1;
+	ShowInfo("storage load complete from DB - id: %d (total: %d)\n", account_id, VECTOR_LENGTH(*p));
+	return num_rows;
```

Comment:
Instead of `num_rows`, this should return `VECTOR_LENGTH(*p)` (they can be different in case there are more items in the SQL database than the `MAX_STORAGE` for example.  You can, in fact, get rid of the `num_rows` variable entirely.